### DISCUSSION TITLE
When to use PA0 or PA_0?

### DISCUSSION THREAD

**Discussion by PerennialNovice** [stm32duino/Arduino_Core_STM32#1954] [Q&A] [answered]
There exit two different definitions for hardware Pins.
It can be written as PA0 or as PA_0.

The PA_0 notation can be used with hardware serial, but it seems to cause some trouble with I²C or SPI instances. Here the PA0 notation has to be used.

Is there some information on what can be used with which APIs?

**Comment by fpistm** [comment, marked as answer]
Hi @MasteringTheMess 
Arduino API uses pin numbers example on uno connector::
- digital pin number: `D0` to `D15`  or 0 to 15
- analog pin number: `A0` to `A5`
For convenience with STM32, we define `PYn` where `Y` is the GPIO port name and `n` the GPIO pin number, example `PA0`.
This pin number is used with all standard Arduino API.

when you used this pin number (`PYn`) the core convert it to the type `PinName`: `PY_n`
This type is a concatenation, ex: `PA_0  = (PortA << 4) + 0x00,` which allows to retrieve the HAL GPIO port and pin.
Some extended API allows to use it for convenience or to go faster, ex: `digitalWriteFast`.
In a general way the pin number is always converted to `PinName` using macros: `digitalPinToPinName`.

**Reply by PerennialNovice** [reply]
Thanks @fpistm for the very abundant explanation :)

so the safe way is to always use the `PYn` notation.
